File: src/emailservice/telemetry.py
# Copyright 2026 Google LLC
# Licensed under the Apache License, Version 2.0 (the "License");

from contextlib import contextmanager

from opentelemetry import propagate, trace
from opentelemetry.trace import SpanKind


_TRACER = trace.get_tracer("online-boutique/messaging")


def _normalized(value):
  return value or "unknown"


def _attributes(subject, kind, envelope):
  return {
      "messaging.system": "nats",
      "messaging.destination.name": subject,
      "messaging.operation.type": kind,
      "messaging.message.id": _normalized(
          getattr(envelope, "message_id", "")),
      "correlation.id": _normalized(
          getattr(envelope, "correlation_id", "")),
  }


@contextmanager
def consumer_span(envelope, subject, kind="event"):
  carrier = {}
  if envelope is not None:
    if envelope.traceparent:
      carrier["traceparent"] = envelope.traceparent
    if envelope.tracestate:
      carrier["tracestate"] = envelope.tracestate
  parent = propagate.extract(carrier)
  with _TRACER.start_as_current_span(
      f"{subject} process",
      context=parent,
      kind=SpanKind.CONSUMER,
      attributes=_attributes(subject, kind, envelope),
  ) as span:
    yield span


def inject_envelope(envelope):
  carrier = {}
  propagate.inject(carrier)
  if carrier.get("traceparent"):
    envelope.traceparent = carrier["traceparent"]
    envelope.tracestate = carrier.get("tracestate", "")
  return envelope

File: src/currencyservice/telemetry.js
/* Copyright 2026 Google LLC. Licensed under the Apache License, Version 2.0. */
'use strict';

const {
  context,
  propagation,
  trace,
  SpanKind,
  SpanStatusCode,
} = require('@opentelemetry/api');

const tracer = trace.getTracer('online-boutique/messaging');

function normalized(value) {
  return value || 'unknown';
}

function messageAttributes({ subject, kind, messageId, correlationId }) {
  return {
    'messaging.system': 'nats',
    'messaging.destination.name': subject,
    'messaging.operation.type': kind,
    'messaging.message.id': normalized(messageId),
    'correlation.id': normalized(correlationId),
  };
}

function extractedContext(envelope = {}) {
  const carrier = {};
  if (envelope.traceparent) carrier.traceparent = envelope.traceparent;
  if (envelope.tracestate) carrier.tracestate = envelope.tracestate;
  return propagation.extract(context.active(), carrier);
}

function recordError(span, error) {
  if (!error) return;
  span.recordException(error);
  span.setStatus({ code: SpanStatusCode.ERROR, message: 'operation failed' });
}

function finishWithinSpan(parentContext, name, options, work) {
  return tracer.startActiveSpan(name, options, parentContext, span => {
    try {
      const result = work(span);
      if (result && typeof result.then === 'function') {
        return result.catch(error => {
          recordError(span, error);
          throw error;
        }).finally(() => span.end());
      }
      span.end();
      return result;
    } catch (error) {
      recordError(span, error);
      span.end();
      throw error;
    }
  });
}

function withConsumerSpan(envelope, options, work) {
  const subject = options.subject;
  return finishWithinSpan(
    extractedContext(envelope),
    `${subject} process`,
    {
      kind: SpanKind.CONSUMER,
      attributes: messageAttributes(options),
    },
    work,
  );
}

function withProducerSpan(options, work) {
  const subject = options.subject;
  return finishWithinSpan(
    context.active(),
    `${subject} publish`,
    {
      kind: SpanKind.PRODUCER,
      attributes: messageAttributes(options),
    },
    work,
  );
}

function injectEnvelope(envelope) {
  const carrier = {};
  propagation.inject(context.active(), carrier);
  if (carrier.traceparent) envelope.traceparent = carrier.traceparent;
  if (carrier.tracestate) envelope.tracestate = carrier.tracestate;
  else if (carrier.traceparent) envelope.tracestate = '';
  return envelope;
}

module.exports = {
  injectEnvelope,
  recordError,
  withConsumerSpan,
  withProducerSpan,
};

